moodle/Sniffs/Commenting/PHPDocTypesSniff.php

```diff
@@ -0,0 +1,960 @@
+
+/**
+ * Check PHPDoc Types.
+ *
+ * @copyright  2024 Otago Polytechnic
+ * @author     James Calder
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later (or CC BY-SA v4 or later)
+ */
+
+namespace MoodleHQ\MoodleCS\moodle\Sniffs\Commenting;
+
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Files\File;
+use MoodleHQ\MoodleCS\moodle\Util\PHPDocTypeParser;
+
+/**
+ * Check PHPDoc Types.
+ */
+class PHPDocTypesSniff implements Sniff
+{
+    /** @var ?File the current file */
+    protected ?File $file = null;
+
+    /** @var array{'code': ?array-key, 'content': string, 'scope_opener'?: int, 'scope_closer'?: int}[]
+     * file tokens */
+    protected array $tokens = [];
+
+    /** @var array<non-empty-string, object{extends: ?non-empty-string, implements: non-empty-string[]}>
+     * classish things: classes, interfaces, traits, and enums */
+    protected array $artifacts = [];
+
+    /** @var ?PHPDocTypeParser */
+    protected ?PHPDocTypeParser $typeparser = null;
+
+    /** @var 1|2 pass 1 for gathering artifact/classish info, 2 for checking */
+    protected int $pass = 1;
+
+    /** @var int current token pointer in the file */
+    protected int $fileptr = 0;
+
+    /** @var non-empty-array<\stdClass&object{type: string, namespace: string, uses: string[], templates: string[],
+     *                                  classname: ?string, parentname: ?string, opened: bool, closer: ?int}>
+     * file scope: classish, function, etc.  We only need a closer if we might be in a switch statement. */
+    protected array $scopes;
+
+    /** @var ?(\stdClass&object{tags: array<string, string[]>}) PHPDoc comment for upcoming declaration */
+    protected ?object $commentpending = null;
+
+    /** @var int how long until we dispose of a pending comment */
+    protected int $commentpendingcounter = 0;
+
+    /** @var ?(\stdClass&object{tags: array<string, string[]>}) PHPDoc comment for current declaration */
+    protected ?object $comment = null;
+
+    /** @var array{'code': ?array-key, 'content': string, 'scope_opener'?: int, 'scope_closer'?: int}
+     * the current token */
+    protected array $token = ['code' => null, 'content' => ''];
+
+    /** @var array{'code': ?array-key, 'content': string, 'scope_opener'?: int, 'scope_closer'?: int}
+     * the previous token */
+    protected array $tokenprevious = ['code' => null, 'content' => ''];
+
+    /**
+     * Register for open tag (only process once per file).
+     * @return array-key[]
+     */
+    public function register(): array {
+        return [T_OPEN_TAG];
+    }
+
+    /**
+     * Processes PHP files and perform PHPDoc type checks with file.
+     * @param File $phpcsfile The file being scanned.
+     * @param int $stackptr The position in the stack.
+     * @return void
+     */
+    public function process(File $phpcsfile, $stackptr): void {
+
+        // Check we haven't already done this file.
+        if ($phpcsfile == $this->file) {
```

Comment:
You're right, it doesn't look like Moodle (4.1 at least) has any PHP files with multiple opening tags.  This is theoretically possible, though, and I would like to handle it, just in case.  Is there a way to tell whether the function's being called to do the checks, or apply the fixes?